### PR TITLE
Py3k port - Initial PR

### PR DESCRIPTION
**Pull request by pythonhacker**
This is pulled from the py3kport branch changes for test_banana.py test with enough changes to do an early review for some core modules. It keeps Py2 compatibility and CI tests should also pass.

Let us not worry too much about the code coverage failures for now as this will be a  WIP and it will change. The main thing is that the CI tests pass.

Briefly the changes are,

```
1. Zope implements -> @implementer decorator and related changes for a few modules.
2. The following modules ported to Python2 compatible 3
   1. banana.py
   2. tokens.py
   3. storage.py
   4. slicers/root.py
   5. slicer.py
```

You can try and look for the places marked with "PY3KPORT". I have tried to be consistent in this in most changes. Thanks.

### CODE REVIEW

**Comment by exarkun** [comment]
Thanks.

> You can try and look for the places marked with "PY3KPORT". I have tried to be consistent in this in most changes. Thanks.

I don't quite follow the meaning of these comments.  Isn't the whole PR Py3k porting changes?  I'm guessing they are meant to convey some extra information but I don't understand what it is.

**Comment by exarkun** [comment]
> You can try and look for the places marked with "PY3KPORT". I have tried to be consistent in this in most changes. Thanks.

For what it's worth, another reviewer might be smart enough to be confident in the changes without test coverage but I doubt I am.  I can review the parts with test coverage but not the other parts.

**Comment by pythonhacker** [comment]
> Thanks.
> 
> > You can try and look for the places marked with "PY3KPORT". I have tried to be consistent in this in most changes. Thanks.
> 
> I don't quite follow the meaning of these comments. Isn't the whole PR Py3k porting changes? I'm > guessing they are meant to convey some extra information but I don't understand what it is.

Meaning, I have indicated non-trivial changes with this comment. The trivial changes would be the ones introduced by 2to3 such as print ... -> print (...) and say map(...) -> list(map(...)) etc. In other words any manual changes I have done. Hope that clarifies.

**Comment by pythonhacker** [comment]
> > Let us not worry too much about the code coverage failures for now
> 
> For what it's worth, another reviewer might be smart enough to be confident in the changes without > test coverage but I doubt I am. I can review the parts with test coverage but not the other parts.

Point is that during a port, you go with the functional changes - that make tests pass and that logically makes sense. Coverage is what I would call as a non-functional architectural quality attribute (from my days as an architect). It can come after the functional port is done. At the moment, if I go and try to find what is happening to the coverage as well, it will be mixing in the functional with the non-functional.

**Comment by pythonhacker** [comment]
I have more changes to push to this branch. Let me know if that is fine or you want me to wait till an initial review is done and/or  (ideally) the current changes are merged.

**Comment by pythonhacker** [comment]
I have more commits to push here - some 14 commits. These changes will make test_banana pass for Py3 while keeping the tox successful for Py2. Will be pushing this as I haven't seen any review comments yet.

**Comment by meejah** [comment]
Just as a general comment, I think this PR is probably in the "too big" category; 900+ changes, 26 commits, 30 files. For future PRs, it would be a LOT easier to review if a single PR was, for example, "change implements to `@implementer` in some modules".

And FWIW, I agree with @exarkun's comments re: test-coverage. It's going to be WAY easier to confirm a change is okay if there are tests to go along with them. (These could also be separate PRs: e.g. one PR that adds a bunch of tests).

File: src/foolscap/schema.py [hunk 1]
```diff
@@ -82,7 +84,10 @@
 DictOf = DictConstraint
 SetOf = SetConstraint
 
-
+if six.PY3:
+    unicode = str
```
Comment (meejah) [inline src/foolscap/schema.py]: This seems a little weird; can we instead just get rid of `unicode()` use entirely..? e.g. by always prefixing strings with `u""` or `b""`?

File: src/foolscap/schema.py [hunk 2]
```diff
@@ -126,23 +131,28 @@ def checkObject(self, obj, inbound):
...
     int: IntegerConstraint(),
     long: IntegerConstraint(maxBytes=1024),
     float: NumberConstraint(),
     None: Nothing(),
     }
 
+if six.PY2:
+    constraintMap[unicode] = UnicodeConstraint()
+    
 # This module provides a function named addToConstraintTypeMap() which helps
```
Comment (meejah) [inline src/foolscap/schema.py]: As above, we should still care about unicode (vs bytes) for py3, no?

File: src/foolscap/slicer.py [hunk 3]
```diff
@@ -129,11 +131,22 @@ def slicerForObject(self, obj):
...
         registry = UnslicerRegistry
-    assert not registry.has_key(opentype)
+    assert opentype not in registry
+
+
     registry[opentype] = factory
 
+    # PY3KPORT  
+    if type(opentype) is tuple:
+        key = (six.ensure_binary(opentype[0]),)
```
Comment (meejah) [inline src/foolscap/slicer.py]: `isinstance` is generally preferred...but also what if this is more than a 1-tuple?

File: src/foolscap/slicers/decimal_slicer.py [hunk 4]
```diff
@@ -33,7 +34,7 @@ def receiveChild(self, obj, ready_deferred=None):
         assert ready_deferred is None
         if self.value != None:
             raise BananaError("already received a string")
-        self.value = decimal.Decimal(obj)
+        self.value = decimal.Decimal(six.ensure_str(obj, 'latin1'))
```
Comment (meejah) [inline src/foolscap/slicers/decimal_slicer.py]: Why `latin1`? (Probably could have a comment, too?)

File: src/foolscap/slicers/dict.py [hunk 5]
```diff
@@ -12,7 +13,7 @@ class DictSlicer(BaseSlicer):
     trackReferences = True
     slices = None
     def sliceBody(self, streamable, banana):
-        for key,value in self.obj.items():
+        for key,value in list(self.obj.items()):
             yield key
```
Comment (meejah) [inline src/foolscap/slicers/dict.py]: shouldn't need `list()` wrapper here.

File: src/foolscap/slicers/dict.py [hunk 6]
```diff
@@ -119,7 +120,7 @@ def describe(self):
 class OrderedDictSlicer(DictSlicer):
     slices = dict
     def sliceBody(self, streamable, banana):
-        keys = self.obj.keys()
+        keys = list(self.obj.keys())
         keys.sort()
         for key in keys:
```
Comment (meejah) [inline src/foolscap/slicers/dict.py]: perhaps `keys = sorted(self.obj.keys())` instead of these three lines?

File: src/foolscap/slicers/root.py [hunk 7]
```diff
@@ -1,7 +1,16 @@
 # -*- test-case-name: foolscap.test.test_banana -*-
 
+try:
+    from types import InstanceType
+except:
+    # PY3KPORT
+    InstanceType = object
+
```
Comment (meejah) [inline src/foolscap/slicers/root.py]: What is going on here?

File: src/foolscap/slicers/root.py [hunk 8]
```diff
@@ -30,6 +40,10 @@ def allowStreaming(self, streamable):
     def registerRefID(self, refid, obj):
         pass
 
+    def next(self):
+        # PY3KPORT
+        return self.__next__()
+
```
Comment (meejah) [inline src/foolscap/slicers/root.py]: See comments higher about "maybe we can refactor a little so we're not recording bound methods into a 3-tuple"...?

File: src/foolscap/slicers/root.py [hunk 9]
```diff
@@ -54,22 +68,56 @@ def slicerForObject(self, obj):
...
+                # MethodSlicer instead!
+                if hasattr(obj, '__qualname__') and '.' in obj.__qualname__:
+                    log.msg("{} is a method, not a function".format(obj))
+                    slicerFactory = self.slicerTable.get(types.MethodType)
+                else:
+                    slicerFactory = self.slicerTable.get(type(obj))
+                    
+            else:
+                slicerFactory = self.slicerTable.get(type(obj))
+
```
Comment (meejah) [inline src/foolscap/slicers/root.py]: I'm not sure I understand what's going on here enough to be convinced we need completely different py2 vs py3 code? (Also this whole added block should go in its own function or method and have unit-tests).

File: src/foolscap/slicers/root.py [hunk 10]
```diff
@@ -54,22 +68,56 @@ def slicerForObject(self, obj):
...
+            else:
+                slicerFactory = self.slicerTable.get(type(obj))
+            
         if slicerFactory:
             if self.debug: log.msg(" got slicerFactory %s" % slicerFactory)
             return slicerFactory(obj)
-        if issubclass(type(obj), types.InstanceType):
+
+        if issubclass(type(obj), InstanceType):
             name = str(obj.__class__)
```
Comment (meejah) [inline src/foolscap/slicers/root.py]: When will a thing NOT be a subclass of `object`...?

File: src/foolscap/slicers/root.py [hunk 11]
```diff
@@ -54,22 +68,56 @@ def slicerForObject(self, obj):
...
 
     def slice(self):
         return self
+    
     def __iter__(self):
         return self # we are our own iterator
-    def next(self):
+
+    def __next__(self):
         if self.objectSentDeferred:
```
Comment (meejah) [inline src/foolscap/slicers/root.py]: wait, don't we alias `__next__` back to `next` above anyway? why the name change?

File: src/foolscap/slicers/root.py [hunk 12]
```diff
@@ -157,7 +205,7 @@ def __init__(self, protocol):
         self.objects = {}
         keys = []
         for r in self.topRegistries + self.openRegistries:
-            for k in r.keys():
+            for k in list(r.keys()):
                 keys.append(len(k[0]))
```
Comment (meejah) [inline src/foolscap/slicers/root.py]: no `list()` wrapper needed.

File: src/foolscap/slicers/unicode.py [hunk 13]
```diff
@@ -1,14 +1,19 @@
...
 
+import six
 import re
 from twisted.internet.defer import Deferred
 from foolscap.tokens import BananaError, STRING, VOCAB, Violation
 from foolscap.slicer import BaseSlicer, LeafUnslicer
 from foolscap.constraint import OpenerConstraint, Any
 
+if six.PY3:
+    unicode = str
```
Comment (meejah) [inline src/foolscap/slicers/unicode.py]: unused?

File: src/foolscap/slicers/vocab.py [hunk 14]
```diff
@@ -32,7 +32,7 @@ def slice(self, streamable, banana):
         stringToIndex = self.obj
         indexToString = dict([(stringToIndex[s],s) for s in stringToIndex])
         assert len(stringToIndex) == len(indexToString) # catch duplicates
-        indices = indexToString.keys()
+        indices = list(indexToString.keys())
         indices.sort()
         for index in indices:
```
Comment (meejah) [inline src/foolscap/slicers/vocab.py]: `for index in sorted(indextoString.keys()):` might be more clear

File: src/foolscap/storage.py [hunk 15]
```diff
@@ -27,7 +28,36 @@
...
+  if inspect.ismethod(meth):
+    # If its a method, we can rely on __self__
+    return meth.__self__.__class__
+
+  if inspect.isfunction(meth):
+    cls = getattr(inspect.getmodule(meth),
+                  meth.__qualname__.split('.<locals>', 1)[0].rsplit('.', 1)[0])
+    if isinstance(cls, type):
+      return cls
+
```
Comment (meejah) [inline src/foolscap/storage.py]: whitespace. and using string-splitting to get .. classes? makes me a little nervous. This needs tests, at minimum.

File: src/foolscap/storage.py [hunk 16]
```diff
@@ -71,9 +101,30 @@ class MethodSlicer(slicer.BaseSlicer):
...
+              yield self.obj.__self__.__class__ 
+          else:
+              # This is an unbound method!
+              # These are very tricky in Python3 as they
+              # are nothing but functions attached to
+              # a namespace (of the class)
+              yield None
+              klass = get_class_that_defined_method(self.obj)
+              # Hack
+              self.obj.im_class = klass
```
Comment (meejah) [inline src/foolscap/storage.py]: yes :) but why?

File: src/foolscap/storage.py [hunk 17]
```diff
@@ -294,14 +354,22 @@ def receiveChild(self, obj, ready_deferred=None):
...
-            assert type(obj) in (types.InstanceType, types.NoneType)
+            if six.PY2:
+              assert type(obj) in (InstanceType, type(None))
+            elif six.PY3:
+              # No more instance types in PY3
+              if hasattr(obj, '__class__'):
+                assert(isinstance(obj.__class__, type))
+              else:
+                assert(type(obj) == type(None))
+
```
Comment (meejah) [inline src/foolscap/storage.py]: if these are critical checks, we shouldn't use `assert` because it isn't run in all cases. e.g. use `if not hasattr(): raise ..` instead.

File: src/foolscap/test/test_banana.py [hunk 18]
```diff
@@ -1,25 +1,87 @@
...
 from twisted.python.failure import Failure
 from twisted.python.components import registerAdapter
 from twisted.internet import defer
 
 from foolscap.tokens import ISlicer, Violation, BananaError
-from foolscap.tokens import BananaFailure, tokenNames, \
-     OPEN, CLOSE, ABORT, INT, LONGINT, NEG, LONGNEG, FLOAT, STRING
+from foolscap.tokens import (BananaFailure, tokenNames, 
+                             OPEN, CLOSE, ABORT, INT, LONGINT, NEG, LONGNEG, FLOAT, STRING)
 from foolscap import slicer, schema, storage, banana, vocab
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: 4-spaces

File: src/foolscap/test/test_banana.py [hunk 19]
```diff
@@ -1,25 +1,87 @@
...
+# PY3KPORT: Some useful utility functions for str<->bytes encoding
+def list_to_bytes(inlist, encoding='latin1'):
+    """ Map each item in the list to bytes wherever possible """
+
+    inlist2 = []
+    for item in inlist:
+        try:
+            inlist2.append(six.ensure_binary(item, encoding))
+        except Exception:
+            inlist2.append(item)
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: `Exception` is too general

File: src/foolscap/test/test_banana.py [hunk 20]
```diff
@@ -1,25 +1,87 @@
...
+
+def dict_key_to_bytes(indict, encoding='latin1'):
+    """ Return a dictionary where keys have been converted from str->byte """
+
+    indict2 = {}
+    for k,v in indict.items():
+        try:
+            indict2[six.ensure_binary(k, encoding)] = v
+        except Exception:
+            indict2[k] = v
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: too general

File: src/foolscap/test/test_banana.py [hunk 21]
```diff
@@ -198,31 +261,36 @@ def shouldDropConnection(self, tokens):
...
-        self.failUnlessEqual(res.where, where)
+        self.assertEqual(res.where, where)
         self.banana.object = None # to stop the tearDown check TODO ??
 
-class TestTransport(StringIO.StringIO):
+if six.PY3:
+    io_klass = six.BytesIO
+else:
+    io_klass = six.StringIO
+
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: Hmm? Why can't we use the same kind of IO on both..?

File: src/foolscap/test/test_banana.py [hunk 22]
```diff
@@ -305,10 +373,14 @@ def shouldDropConnection(self, stream):
...
 
     def wantEqual(self, got, wanted):
-        if got != wanted:
-            print
-            print "wanted: '%s'" % wanted, repr(wanted)
-            print "got   : '%s'" % got, repr(got)
+
+        got_b = six.ensure_binary(got, 'latin1')
+        wanted_b = six.ensure_binary(wanted, 'latin1')
+
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: Can you explain why the user-code (i.e. tests) need to do this conversion?

File: src/foolscap/test/test_banana.py [hunk 23]
```diff
@@ -323,9 +395,17 @@ def looptest(self, obj, newvalue=_None):
...
-        self.failUnlessEqual(type(obj2), type(newvalue))
+        if type(obj2) is dict:
+            self.assertEqual(dict_key_to_bytes(obj2),
+                             dict_key_to_bytes(newvalue))
+        elif type(obj2) in (str, bytes):
+            self.assertEqual(six.ensure_binary(obj2, 'latin1'),
+                             six.ensure_binary(newvalue, 'latin1'))
+        else:
+            self.assertEqual(obj2, newvalue)
+            self.assertEqual(type(obj2), type(newvalue))
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: same as above; why do the tests need to do this conversion?

File: src/foolscap/test/test_banana.py [hunk 24]
```diff
@@ -398,60 +478,63 @@ def test_aborted_list3(self):
...
         f = self.shouldDropConnection([tOPEN(0),'dict',
                                        "a",1,"a",2,
                                        tCLOSE(0)])
-        self.failUnlessEqual(f.value.where, "<RootUnslicer>.{}")
-        self.failUnlessEqual(f.value.args[0], "duplicate key 'a'")
+        self.assertEqual(f.value.where, "<RootUnslicer>.{}")
+        if six.PY2:
+            self.assertEqual(f.value.args[0], "duplicate key 'a'")
+        elif six.PY3:
+            self.assertEqual(f.value.args[0], "duplicate key 'b'a''")
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: Instead it's probably better to ensure we give consistent error-message across py2/3

File: src/foolscap/test/test_banana.py [hunk 25]
```diff
@@ -476,8 +559,12 @@ def test_instance_bad1(self):
...
         f = self.shouldDropConnection(tokens)
-        self.failUnlessEqual(f.value.where, "<RootUnslicer>.<Foo>.c.<??>")
-        self.failUnlessEqual(f.value.args[0],
+        if six.PY2:
+            self.assertEqual(f.value.where, "<RootUnslicer>.<Foo>.c.<??>")
+        elif six.PY3:
+            self.assertEqual(f.value.where, "<RootUnslicer>.<b'Foo'>.b'c'.<??>")
+
+        self.assertEqual(f.value.args[0],
                              "InstanceUnslicer classname must be string")
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: better to have consistent error-messages instead of special-casing the tests.

File: src/foolscap/test/test_banana.py [hunk 26]
```diff
@@ -491,9 +578,14 @@ def test_instance_bad2(self):
...
-        self.failUnlessEqual(f.value.args[0],
+        if six.PY2:
+            self.assertEqual(f.value.where,
+                            "<RootUnslicer>.<Foo>.c.<Bar>.attrname??")
+        else:
+            self.assertEqual(f.value.where,
+                            "<RootUnslicer>.<b'Foo'>.b'c'.<b'Bar'>.attrname??")
+
+        self.assertEqual(f.value.args[0],
                              "InstanceUnslicer keys must be STRINGs")
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: ditto

File: src/foolscap/test/test_banana.py [hunk 27]
```diff
@@ -511,17 +603,20 @@ def test_instance_unsafe1(self):
...
         f = self.shouldFail(tokens)
-        self.failUnlessEqual(f.value.where, "<RootUnslicer>")
-        self.failUnlessEqual(f.value.args[0],
-                             "unknown top-level OPEN type ('instance',)")
+        self.assertEqual(f.value.where, "<RootUnslicer>")
+        if six.PY2:
+            self.assertEqual(f.value.args[0], "unknown top-level OPEN type ('instance',)")
+        elif six.PY3:
+            self.assertEqual(f.value.args[0], "unknown top-level OPEN type (b'instance',)")         
+
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: ditto

File: src/foolscap/test/test_banana.py [hunk 28]
```diff
@@ -618,33 +713,36 @@ def test_failed_dict1(self):
...
-        self.failUnless(f.check(Violation))
-        self.failUnlessEqual(f.value.where, "<RootUnslicer>.[1]")
-        self.failUnlessEqual(f.value.args[0], "unknown OPEN type ('bad',)")
+        self.assertTrue(isinstance(f, BananaFailure))
+        self.assertTrue(f.check(Violation))
+        self.assertEqual(f.value.where, "<RootUnslicer>.[1]")
+        if six.PY2:
+            self.assertEqual(f.value.args[0], "unknown OPEN type ('bad',)")
+        elif six.PY3:
+            self.assertEqual(f.value.args[0], "unknown OPEN type (b'bad',)")
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: ditto

File: src/foolscap/test/test_banana.py [hunk 29]
```diff
@@ -654,10 +752,14 @@ def test_failed_dict4(self):
...
-        self.failUnlessEqual(f.value.where, "<RootUnslicer>.[1].{}[b]")
-        self.failUnlessEqual(f.value.args[0], "aaaaaaaaargh")
+        self.assertTrue(isinstance(f, BananaFailure))
+        self.assertTrue(f.check(Violation))
+        if six.PY2:
+            self.assertEqual(f.value.where, "<RootUnslicer>.[1].{}[b]")
+        elif six.PY3:
+            self.assertEqual(f.value.where, "<RootUnslicer>.[1].{}[b'b']")
+
+        self.assertEqual(f.value.args[0], "aaaaaaaaargh")
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: consistent error-messages instead

File: src/foolscap/test/test_banana.py [hunk 30]
```diff
@@ -667,17 +769,39 @@ def test_failed_dict5(self):
...
-        return cmp((self.__class__, self.__dict__),
-                   (them.__class__, them.__dict__))
+
+        if six.PY2:
+            return cmp((self.__class__, self.__dict__),
+                       (them.__class__, them.__dict__))
+        elif six.PY3:
+            # PY3KPORT
+            # Can have bytes as keys from strings
+            self_bdict = dict_key_to_bytes(self.__dict__)
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: What does this mean? Do we have inconsistent key-types between py2 and 3?

File: src/foolscap/test/test_banana.py [hunk 31]
```diff
@@ -855,6 +979,9 @@ def slice(self, streamable, banana):
         return self
 
     def next(self):
+        return self.__next__()
+
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: Why do we need this alias? (e.g. if we really do, explain it in the docstring)

File: src/foolscap/test/test_banana.py [hunk 32]
```diff
@@ -975,21 +1102,25 @@ def test4(self):
...
+        self.assertEqual(e.value.where, "<RootSlicer>.ErrorfulSlicer[1]")
+        if six.PY2:
+            self.failUnlessSubstring("cannot serialize <open file",
+                                     e.value.args[0])
+        elif six.PY3:
+            self.failUnlessSubstring("cannot serialize <_io.TextIOWrapper",
+                                     e.value.args[0])       
+        self.assertTrue(s.childDied)
+        self.assertEqual(self.banana.tokens,
                              [('OPEN', 0), 1, ('ABORT',), ('CLOSE', 0)])
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: Can we have consistent error-messages instead?

File: src/foolscap/test/test_banana.py [hunk 33]
```diff
@@ -1326,7 +1459,18 @@ def check(self, obj, stream):
...
+            if type(obj) is dict:
+                obj = dict_value_to_bytes(obj)
+                obj2 = dict_value_to_bytes(obj2)
+            elif type(obj) is list:
+                obj = list_to_bytes(obj)
+                obj2 = list_to_bytes(obj2)
+            elif type(obj) in (str, bytes):
+                obj = six.ensure_binary(obj)
+                obj2 = six.ensure_binary(obj2)
+
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: This all makes be nervous. I don't know why it's needed.

File: src/foolscap/test/test_banana.py [hunk 34]
```diff
@@ -1418,13 +1562,28 @@ def setConstraints(self, constraint, childConstraint):
...
+            if type(obj) is dict:
+                obj = dict_value_to_bytes(obj)
+                obj2 = dict_value_to_bytes(obj2)
+            elif type(obj) is list:
+                obj = list_to_bytes(obj)
+                obj2 = list_to_bytes(obj2)
+            elif type(obj) is tuple:
+                obj = tuple(list_to_bytes(obj))
+                obj2 = tuple(list_to_bytes(obj2))
+
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: same; feeling nervous ;)

File: src/foolscap/test/test_banana.py [hunk 35]
```diff
@@ -1627,18 +1786,17 @@ def testConstrainedBool(self):
                       schema.BooleanConstraint(True))
 
 
-class A:
+# PY3KPORT: Let these test classes inherit from Bar as it
+# already contains our Py2 & 3 portable code for comparison/equality
+# Makes the code simpler.
+class A(Bar):
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: ...but it also means there no "old-style" class in the test-suite now, right?

File: src/foolscap/test/test_banana.py [hunk 36]
```diff
@@ -1691,14 +1849,14 @@ def test_decimal(self):
...
+            self.assertEqual(str(new_NaN), str(Decimal('NaN')))
         d.addCallback(_check_NaN)
         return d
 
     def test_string(self):
         return self.looptest("biggles")
     def test_unicode(self):
-        return self.looptest(u"biggles\u1234")
+        return self.looptest("biggles\\u1234")
     def test_list(self):
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: this will now be `str` and not `unicode` on py2, no? and the double-slash means no special unicode char?

File: src/foolscap/test/test_banana.py [hunk 37]
```diff
@@ -1787,13 +1945,21 @@ def _testIdentity_1(self, z):
...
-        self.assertEquals(x, y)
-        self.assertEquals(type(x[0]), type(y[0]))
+
+        if type(x) in (str, bytes):
+            x = six.ensure_text(x)
+            y = six.ensure_text(y)
+        elif type(x) in (list, tuple):
+            x = list(map(six.ensure_text, x))
+            y = list(map(six.ensure_text, y))
+
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: should use list-comprehensions, not `map` in new code

File: src/foolscap/test/test_banana.py [hunk 38]
```diff
@@ -1878,7 +2044,7 @@ def test_incoming1(self):
         b = TokenBanana()
         b.connectionMade()
         vdict = {1: 'list', 2: 'tuple', 3: 'dict'}
-        keys = vdict.keys()
+        keys = list(vdict.keys())
         keys.sort()
         setVdict = [tOPEN(0),'set-vocab']
         for k in keys:
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: `for k in sorted(vdict.keys()):` perhaps

File: src/foolscap/test/test_banana.py [hunk 39]
```diff
@@ -1887,15 +2053,15 @@ def test_incoming1(self):
...
     def test_outgoing(self):
         b = TokenBanana()
         b.connectionMade()
         b.tokens = []
         strings = ["list", "tuple", "dict"]
         vdict = {0: 'list', 1: 'tuple', 2: 'dict'}
-        keys = vdict.keys()
+        keys = list(vdict.keys())
         keys.sort()
         setVdict = [tOPEN(0),'set-vocab']
```
Comment (meejah) [inline src/foolscap/test/test_banana.py]: same `sorted(..)` comment

File: src/foolscap/banana.py [hunk 40]
```diff
@@ -10,31 +11,47 @@
...
-from tokens import SIZE_LIMIT, STRING, LIST, INT, NEG, \
-     LONGINT, LONGNEG, VOCAB, FLOAT, OPEN, CLOSE, ABORT, ERROR, \
-     PING, PONG, \
-     BananaError, BananaFailure, Violation
+from . import stringchain
+from . import tokens
+from .tokens import (SIZE_LIMIT, STRING, LIST, INT, NEG, 
+                    LONGINT, LONGNEG, VOCAB, FLOAT, OPEN, CLOSE, ABORT, ERROR, 
+                    PING, PONG, 
+                    BananaError, BananaFailure, Violation)
```
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: > Okay, after churning through this monster I'd like to re-iterate my previous general comment: this is really too big and too many different kinds of changes to review properly. It would be way better if, e.g., you had a single PR for "failUnless -> assert.." changes, one for "six.bytes" related changes, etc etc. because then we can merge some things faster.
> 
> As it stands, this needs changes in some spots .. but I'll have to re-read through all the "simple/mechanical" changes again every time ... :/

I understand the spirit of this but for a port and to keep the tests passing with the changes you understand what you are saying is difficult. 

I can rework this to multiple changes  like you requested from next PR - so pls confirm if you want the current PR to be reworked or if its fine from the next one. Thanks.

File: src/foolscap/banana.py [hunk 41]
```diff
@@ -10,31 +11,47 @@
...
+
+def ensure_byte(ch):
+    # PY3PORT - new function
+    if six.PY2:
+        return ch
+    elif six.PY3:
+        return six.int2byte(ch)
+
 def int2b128(integer, stream):
     if integer == 0:
```
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: int2b128 did not have docstring already. Since its a port, I wont be adding docstrings for entities which did not have it already for keeping the changes efficient. For any new functions I add, yes, I will be adding docstrings.

File: src/foolscap/banana.py [hunk 42]
```diff
@@ -10,31 +11,47 @@
...
-        stream(chr(0))
+        # PY3KPORT
+        stream(six.int2byte(0))
         return
     assert integer > 0, "can only encode positive integers"
     while integer:
-        stream(chr(integer & 0x7f))
+        # PY3KPORT      
+        stream(six.int2byte(integer & 0x7f))
```
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: I can remove them right away. I thought it helps, looks like it does not so I am going to remove them, stepwise in every module with further changes.

File: src/foolscap/test/test_banana.py (hunk 35 above)
Comment (pythonhacker) [reply src/foolscap/test/test_banana.py]: Nope. Class A inherits from Bar which is an old style class in Python2, so there are still old style classes in the test suite. 

```
>>> type(test_banana.A)
<type 'classobj'>
>>> type(test_banana.B)
<type 'classobj'>
```

File: src/foolscap/test/test_banana.py (hunk 37 above)
Comment (pythonhacker) [reply src/foolscap/test/test_banana.py]: Changed. Thank you.

File: src/foolscap/test/test_banana.py (hunk 38 above)
Comment (pythonhacker) [reply src/foolscap/test/test_banana.py]: Yes. Makes sense. Changed.

File: src/foolscap/test/test_banana.py (hunk 39 above)
Comment (pythonhacker) [reply src/foolscap/test/test_banana.py]: Changed.

File: src/foolscap/banana.py (hunk 42 above)
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: Changed for this module (test_banana.py)

File: src/foolscap/banana.py [hunk 43]
```diff
@@ -10,31 +11,47 @@
...
+                    BananaError, BananaFailure, Violation)
 
 EPSILON = 0.1
 
+if six.PY3:
+    long = int
+
+def ensure_byte(ch):
+    # PY3PORT - new function
+    if six.PY2:
```
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: Added.

File: src/foolscap/test/test_banana.py (hunk 36 above)
Comment (pythonhacker) [reply src/foolscap/test/test_banana.py]: This was perhaps a wrong escape added by 2to3. Checking.

File: src/foolscap/banana.py (hunk 40 above)
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: > I don't know what the convention (if any?) in foolscap is, but in Tahoe we'd typically indent 
> follow-on lines at 4 spaces.
Personally I tend to align the starting character of each module to same column, so that is what I have done here. And I use tab-to-spaces in my Emacs so any final code will have tabs converted to spaces. Just saying. Thanks.

[the remark this answers is not in the document]
Comment (meejah) [reply src/foolscap/banana.py]: > I can rework this to multiple changes like you requested from next PR - so pls confirm if you want the current PR to be reworked or if its fine from the next one. Thanks.

I think there are lots of changes in here that would keep tests passing .. e.g. any whitespace changes, any of the test-assertion-function changes, all the `print` -> `print()` changes, using `six.bytes` (or `b""` / `u""`) etc. I guess it's up to you for this PR, but those would make some fine "smaller" PRs and could be easily merged.

[the remark this answers is not in the document]
Comment (meejah) [reply src/foolscap/banana.py]: > For imports, I tend to align the starting character of each module to same column for multi-line imports,

Okay, but it seems the convention is to use 4-space indenting already on continuation lines in Foolscap?

File: src/foolscap/banana.py (hunk 41 above)
[the remark this answers is not in the document]
Comment (meejah) [reply src/foolscap/banana.py]: I think functions that get changed can benefit from a docstring, too. Then reviewers don't have to (also) convince themselves what the function is supposed to be doing..Obviously, we have to limit scope somewhere (can't fix everything at once), but...

File: src/foolscap/banana.py (hunk 40 above)
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: Ok. I can rework this into multiple PRs. Let me do that. First one would be the one with Zope changes.

File: src/foolscap/constraint.py [hunk 44]
```diff
@@ -5,12 +5,13 @@
...
 import re
-from zope.interface import implements, Interface
+from zope.interface import implementer, Interface
 
-from foolscap.tokens import Violation, BananaError, SIZE_LIMIT, \
-     STRING, LIST, INT, NEG, LONGINT, LONGNEG, VOCAB, FLOAT, OPEN, \
-     tokenNames
+from foolscap.tokens import (Violation, BananaError, SIZE_LIMIT, 
+                             STRING, LIST, INT, NEG, LONGINT, LONGNEG, VOCAB, FLOAT, OPEN, 
+                             tokenNames)
```
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/constraint.py]: Ok. In the PR for this.

File: src/foolscap/schema.py (hunk 1 above)
Comment (pythonhacker) [reply src/foolscap/schema.py]: We can use "six.ensure_text(..)" etc for this. In this case I went for a quick shortcut really.

File: src/foolscap/slicers/decimal_slicer.py (hunk 4 above)
Comment (pythonhacker) [reply src/foolscap/slicers/decimal_slicer.py]: Default is utf-8. In this case these are always number strings like "10000" so latin1 encoding is kind of assured. Can be left to default as well. Dont remember which context I did it specifically.

**Comment by pythonhacker** [comment]
A partial PR with more incremental only changes at https://github.com/warner/foolscap/pull/53 . I will keep this open for a while for any more comments but request to review the other one so we can merge changes here in an incremental way as suggested by @meejah  .

File: src/foolscap/banana.py (hunk 40 above)
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: See https://github.com/warner/foolscap/pull/53 . Thanks.

File: src/foolscap/test/test_banana.py (hunk 36 above)
Comment (pythonhacker) [reply src/foolscap/test/test_banana.py]: Fixed.

File: src/foolscap/banana.py [hunk 45]
```diff
@@ -10,31 +11,47 @@
...
         return
     assert integer > 0, "can only encode positive integers"
     while integer:
-        stream(chr(integer & 0x7f))
+        # PY3KPORT      
+        stream(six.int2byte(integer & 0x7f))
         integer = integer >> 7
-
+        
 def b1282int(st):
```
[the remark this answers is not in the document]
Comment (pythonhacker) [reply src/foolscap/banana.py]: Fixed (I guess).

**Comment by pythonhacker** [comment]
I am using the review feedback here to make code changes. Do not merge this, will close this by end of this week. Thanks.

**Comment by pythonhacker** [comment]
Closing this.
